cpp/include/cuspatial/experimental/detail/join/intersection.cuh

```diff
@@ -0,0 +1,146 @@
+                                                      IsInternalIterator is_internal_node_first,
+                                                      BoundingBoxIterator bounding_box_first,
+                                                      NodeIndicesIterator node_indices,
+                                                      BBoxIndicesIterator bbox_indices,
+                                                      NodePairsIterator node_pairs,
+                                                      LeafPairsIterator leaf_pairs,
+                                                      int32_t num_pairs,
+                                                      T x_min,
+                                                      T y_min,
+                                                      T scale,
+                                                      int8_t max_depth,
+                                                      rmm::cuda_stream_view stream)
+{
+  auto nodes_first = thrust::make_zip_iterator(keys_first, levels_first, is_internal_node_first);
+
+  thrust::transform(
+    rmm::exec_policy(stream),
+    thrust::make_zip_iterator(node_indices, bbox_indices),
+    thrust::make_zip_iterator(node_indices, bbox_indices) + num_pairs,
+    node_pairs,
+    [x_min, y_min, scale, max_depth, nodes = nodes_first, bboxes = bounding_box_first] __device__(
+      auto const& node_and_bbox) {
+      auto const& node_idx = thrust::get<0>(node_and_bbox);
+      auto const& bbox_idx = thrust::get<1>(node_and_bbox);
+
+      auto const& node                = nodes[node_idx];
+      uint32_t const& key             = thrust::get<0>(node);
+      uint8_t const& level            = thrust::get<1>(node);
+      uint8_t const& is_internal_node = thrust::get<2>(node);
+
+      auto const& bbox       = bboxes[bbox_idx];
+      auto const& bbox_min   = bbox.v1;
+      auto const& bbox_max   = bbox.v2;
+      auto const& poly_x_min = bbox_min.x;
+      auto const& poly_y_min = bbox_min.y;
+      auto const& poly_x_max = bbox_max.x;
+      auto const& poly_y_max = bbox_max.y;
+
+      T const key_x       = utility::z_order_x(key);
+      T const key_y       = utility::z_order_y(key);
+      T const level_scale = scale * (1 << (max_depth - 1 - level));
+      T const node_x_min  = x_min + (key_x + 0) * level_scale;
+      T const node_y_min  = y_min + (key_y + 0) * level_scale;
+      T const node_x_max  = x_min + (key_x + 1) * level_scale;
+      T const node_y_max  = y_min + (key_y + 1) * level_scale;
+
+      if ((node_x_min > poly_x_max) || (node_x_max < poly_x_min) || (node_y_min > poly_y_max) ||
+          (node_y_max < poly_y_min)) {
+        // if no overlap, return type = none_indicator
+        return thrust::make_tuple(none_indicator, level, node_idx, bbox_idx);
+      }
+      // otherwise return type = leaf_indicator (0) or quad_indicator (1)
+      return thrust::make_tuple(is_internal_node, level, node_idx, bbox_idx);
```

Comment:
It would be nice to have a `box.overlap()` helper that we can just call 
```c++
return bbox.overlaps(node_bbox) ? tuple(is_internal_node, ..) : tuple(none_indicator, ..);
```
This is a really high level thought, so no need to implement.